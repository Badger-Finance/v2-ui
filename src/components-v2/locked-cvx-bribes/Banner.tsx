import React, { useContext } from 'react';
import { Card, Grid, Typography } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import Delegation from './Delegation';
import Earnings from './Earnings';
import { observer } from 'mobx-react-lite';
import { StoreContext } from '../../mobx/store-context';

export const SAFE_BOX_ILLUSTRATION_BREAKPOINT = 1350;

const useStyles = makeStyles((theme) => ({
	root: {
		marginTop: theme.spacing(4),
		padding: theme.spacing(4),
		backgroundColor: '#3A3A3A',
		position: 'relative',
	},
	delegationSection: {
		padding: theme.spacing(3),
	},
	title: {
		fontSize: 32,
		fontWeight: 'bolder',
		marginBottom: theme.spacing(2),
	},
	descriptionSection: {
		marginBottom: theme.spacing(3),
	},
	safeLockImageLg: {
		position: 'absolute',
		bottom: 10,
		right: -10,
		height: 220,
		[theme.breakpoints.down(SAFE_BOX_ILLUSTRATION_BREAKPOINT)]: {
			display: 'none',
		},
	},
	safeLockImageSm: {
		margin: theme.spacing(2, 0, 4, 4),
		width: '100%',
		[theme.breakpoints.up('sm')]: {
			display: 'none',
		},
	},
	blueParticle: {
		width: 11.4,
		height: 11.4,
		backgroundColor: '#1682FE',
		position: 'absolute',
		top: 27,
		left: 0,
		[theme.breakpoints.down('xs')]: {
			display: 'none',
		},
	},
	greenParticle: {
		width: 11.4,
		height: 11.4,
		backgroundColor: '#60D8A4',
		position: 'absolute',
		top: 56.4,
		left: '50%',
		[theme.breakpoints.down('md')]: {
			display: 'none',
		},
	},
	diagonalParticles: {
		position: 'absolute',
		top: 29,
		left: '80%',
		[theme.breakpoints.down('xs')]: {
			display: 'none',
		},
	},
	descriptionText: {
		fontWeight: 400,
	},
}));

const Banner = (): JSX.Element | null => {
	const {
		lockedCvxDelegation: { shouldBannerBeDisplayed },
	} = useContext(StoreContext);

	const classes = useStyles();

	if (!shouldBannerBeDisplayed) {
		return null;
	}

	return (
		<Card classes={{ root: classes.root }}>
			<Typography className={classes.title}>Delegate to Earn Rewards</Typography>
			<div className={classes.blueParticle} />
			<div className={classes.greenParticle} />
			<img src="assets/diagonal-particles.svg" className={classes.diagonalParticles} alt="diagonal-particles" />
			<Grid container spacing={3}>
				<Grid container item xs={12} sm>
					<img
						src="assets/locked-cvx-safebox-sm.png"
						className={classes.safeLockImageSm}
						alt="cvx locked safebox illustration"
					/>
					<div className={classes.descriptionSection}>
						<Typography variant="body1" className={classes.descriptionText}>
							Earn a portion of the Votium incentives by simply delegating your Vote Locked Convex to
							Badger today.
						</Typography>
					</div>
					<Grid container component={Card} className={classes.delegationSection}>
						<Delegation />
					</Grid>
				</Grid>
				<Grid item xs={12} sm>
					<Earnings />
				</Grid>
			</Grid>
			<img src="assets/locked-cvx-safebox-lg.png" className={classes.safeLockImageLg} alt="safe-lock-image" />
		</Card>
	);
};

export default observer(Banner);
