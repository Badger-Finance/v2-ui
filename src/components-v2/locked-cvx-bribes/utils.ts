import BigNumber from 'bignumber.js';
import { formatWithoutExtraZeros, numberWithCommas, unscale } from '../../mobx/utils/helpers';

export function formatBalance(balance: BigNumber): string {
  return numberWithCommas(formatWithoutExtraZeros(unscale(balance, 18)));
}
