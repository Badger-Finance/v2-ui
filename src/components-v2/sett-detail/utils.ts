import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/setts/sett-charts';
import { DelaySeverity } from '../../mobx/model/setts/sett-rewards';
import { SettDetailMode } from '../../mobx/model/setts/sett-detail';

dayjs.extend(utc);

export const ChartModeTitles = {
	[ChartMode.value]: 'Sett Value',
	[ChartMode.ratio]: 'Token Ratio',
	[ChartMode.accountBalance]: 'Your Holdings',
};

export const SettModeTitles = {
	[SettDetailMode.settInformation]: 'Vault Info',
	[SettDetailMode.userInformation]: 'User Info',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};
