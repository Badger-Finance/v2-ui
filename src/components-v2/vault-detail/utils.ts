import { Vault, VaultData } from '@badger-dao/sdk';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/vaults/vault-charts';
import { DelaySeverity } from '../../mobx/model/vaults/vault-rewards';

dayjs.extend(utc);

export const ChartModeTitles: Record<string, string> = {
	[ChartMode.Value]: 'Sett Value',
	[ChartMode.Ratio]: 'Token Ratio',
	[ChartMode.AccountBalance]: 'My Holdings',
	[ChartMode.BoostMultiplier]: 'Badger Boost',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};

export function defaultVaultBalance(vault: Vault): VaultData {
	return {
		address: vault.vaultToken,
		name: vault.name,
		symbol: vault.asset,
		pricePerFullShare: vault.pricePerFullShare,
		balance: 0,
		value: 0,
		earnedBalance: 0,
		earnedValue: 0,
		depositedBalance: 0,
		withdrawnBalance: 0,
		tokens: [],
		earnedTokens: [],
	};
}

export interface Balance {
	balance: number;
}

export function hasBalance(input?: Balance): boolean {
	if (!input) {
		return false;
	}
	return input.balance > 0;
}
