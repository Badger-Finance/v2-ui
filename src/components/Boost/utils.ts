import { makeStyles } from '@material-ui/core/styles';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import React from 'react';

type ComparisonConfig = {
  toCompareValue: number;
  toBeComparedValue: number;
  greaterCaseColor: string;
  lessCaseColor: string;
  defaultColor: string;
};

export const getColorFromComparison = ({
  toCompareValue,
  toBeComparedValue,
  greaterCaseColor,
  lessCaseColor,
  defaultColor,
}: ComparisonConfig): string => {
  if (toCompareValue > toBeComparedValue) return greaterCaseColor;
  if (toCompareValue < toBeComparedValue) return lessCaseColor;

  return defaultColor;
};

export const useAssetInputStyles = (
  currentValue: string,
  toCompare = 0,
): ((props?: React.CSSProperties) => ClassNameMap<'assetColor'>) => {
  return makeStyles((theme) => {
    const defaultColor = currentValue ? theme.palette.text.primary : theme.palette.text.secondary;
    const fontColor = getColorFromComparison({
      toCompareValue: Number(currentValue),
      toBeComparedValue: toCompare,
      greaterCaseColor: '#74D189',
      lessCaseColor: theme.palette.error.main,
      defaultColor,
    });

    return {
      assetColor: {
        color: fontColor,
      },
    };
  });
};
