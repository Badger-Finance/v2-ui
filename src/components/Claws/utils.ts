import BigNumber from 'bignumber.js';
import { INVALID_REASON } from './index';

export interface ValidationParams {
	amount: string | BigNumber;
	maximum: BigNumber;
	minimum?: BigNumber;
}

export function validateAmountBoundaries({
	amount,
	maximum,
	minimum = new BigNumber('0'),
}: ValidationParams): INVALID_REASON | undefined {
	let error: INVALID_REASON | undefined = undefined;

	const input = new BigNumber(amount);
	const amountExceedsBalance = input.gt(maximum);
	const amountIsLessThanMinimum = input.lt(minimum);

	if (amountIsLessThanMinimum) error = INVALID_REASON.UNDER_MINIMUM;
	if (amountExceedsBalance) error = INVALID_REASON.OVER_MAXIMUM;

	console.log({
		input: input.toString(),
		minimum: minimum.toString(),
		maximum: maximum.toString(),
		amountExceedsBalance,
		amountIsLessThanMinimum,
		error,
	});

	return error;
}
