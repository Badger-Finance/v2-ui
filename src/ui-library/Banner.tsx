import { Grid, Link, makeStyles, Typography } from '@material-ui/core';
import React from 'react';

export interface BannerProps {
  message: string;
  linkText?: string;
  link?: string;
  closeElement?: React.ReactNode;
}

const useStyles = makeStyles({
  root: {
    minHeight: 48,
    padding: '14px 26px',
    background: '#BDE0FF',
  },
  messageLinkContainer: {
    display: 'flex',
  },
  linkIcon: {
    marginRight: 6,
  },
  actionButton: {
    color: '#2E44C0',
  },
  actionLabel: {
    fontSize: 14,
    fontWeight: 500,
    letterSpacing: 1.25,
    textTransform: 'uppercase',
  },
  link: {
    color: '#2E44C0',
    letterSpacing: '0.0025em',
    fontSize: 14,
    fontWeight: 'bold',
  },
  message: {
    fontWeight: 'normal',
    fontSize: 14,
    letterSpacing: 0.25,
    color: 'rgba(0, 0, 0, 0.87)',
    marginRight: 12,
  },
  actionContainer: {
    paddingLeft: 40,
  },
  closeIcon: {
    color: '#2E44C0',
    margin: '-12px',
    '& svg': {
      fontSize: 24,
    },
  },
});

const Banner = ({ message, link, linkText, closeElement = null }: BannerProps): JSX.Element => {
  const classes = useStyles();
  return (
    <Grid container alignItems="center" className={classes.root} justifyContent="space-between">
      <Grid item container xs alignItems="center" className={classes.messageLinkContainer}>
        <Grid item>
          <Typography className={classes.message}>{message}</Typography>
        </Grid>
        {link && linkText && (
          <Grid item>
            <Link href={link} rel="noreferrer" target="_blank" className={classes.link}>
              <img className={classes.linkIcon} src="/assets/icons/banner-link-icon.svg" alt="banner link icon" />
              {linkText}
            </Link>
          </Grid>
        )}
      </Grid>
      {!!closeElement && (
        <Grid item xs="auto" className={classes.actionContainer}>
          {closeElement}
        </Grid>
      )}
    </Grid>
  );
};

export default Banner;
