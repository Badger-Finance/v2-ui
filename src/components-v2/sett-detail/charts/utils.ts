import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { Sett } from '../../../mobx/model/setts/sett';
import { fetchSettChartInformation } from '../../../mobx/utils/apiV2';
import { SettSnapshotGranularity } from '../../../mobx/model/setts/sett-snapshot';
import { SettChartData } from '../../../mobx/model/setts/sett-charts';

dayjs.extend(utc);

export enum SettChartTimeframe {
	'day' = 1,
	'week' = 7,
	'month' = 30,
}

/**
 * Fetches chart information since the provided timeframe until the current date
 * @param sett
 * @param timeframe
 */
export const fetchSettChart = async (sett: Sett, timeframe: SettChartTimeframe): Promise<SettChartData[] | null> => {
	const timeframeDays = 1 * timeframe;
	const isDayTimeFrame = timeframe === SettChartTimeframe.day;

	// if timeframe is just one day then we want the granularity to be hours
	const granularity = isDayTimeFrame ? SettSnapshotGranularity.HOUR : SettSnapshotGranularity.DAY;

	const now = dayjs().utc(); // query until current date
	const from = dayjs(now).subtract(timeframeDays, 'days').utc();

	const fetchedData = await fetchSettChartInformation(sett.vaultToken, from.toDate(), now.toDate(), granularity);

	if (!fetchedData) {
		return null;
	}

	// data needs to be ascending sorted
	return fetchedData
		.sort((a, b) => a.timestamp - b.timestamp)
		.map((d) => ({ ...d, timestamp: new Date(d.timestamp) }));
};
