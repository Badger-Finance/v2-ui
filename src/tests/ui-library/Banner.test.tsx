import React from 'react';
import { checkSnapshot } from '../utils/snapshots';
import Banner from '../../ui-library/Banner';
import { customRender, fireEvent, screen } from '../Utils';
import BannerCloseActionButton from '../../ui-library/BannerCloseActionButton';

describe('Banner', () => {
	it('renders banner information correctly', () => {
		checkSnapshot(<Banner message={'This is an important message'} />);
	});

	it('uses correct link', () => {
		checkSnapshot(
			<Banner message={'This is an important message'} link="https://badger.com" linkText="view more" />,
		);
	});

	it('can trigger close action', () => {
		const closeMock = jest.fn();
		customRender(
			<Banner
				message={'This is an important message'}
				link="https://badger.com"
				linkText="view more"
				closeElement={<BannerCloseActionButton text="dismiss" onClose={closeMock} />}
			/>,
		);
		fireEvent.click(screen.getByRole('button', { name: 'dismiss', exact: false }));
		expect(closeMock).toHaveBeenCalledTimes(1);
	});
});
