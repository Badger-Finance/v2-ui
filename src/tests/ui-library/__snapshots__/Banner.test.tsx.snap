// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner renders banner information correctly 1`] = `
<div>
  <div
    class="MuiGrid-root-10 makeStyles-root-1 MuiGrid-container-11 MuiGrid-align-items-xs-center-19 MuiGrid-justify-content-xs-space-between-30"
  >
    <div
      class="MuiGrid-root-10 makeStyles-messageLinkContainer-2 MuiGrid-container-11 MuiGrid-item-12 MuiGrid-align-items-xs-center-19 MuiGrid-grid-xs-true-44"
    >
      <div
        class="MuiGrid-root-10 MuiGrid-item-12"
      >
        <p
          class="MuiTypography-root-113 makeStyles-message-7 MuiTypography-body1-115"
        >
          This is an important message
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Banner uses correct link 1`] = `
<div>
  <div
    class="MuiGrid-root-10 makeStyles-root-1 MuiGrid-container-11 MuiGrid-align-items-xs-center-19 MuiGrid-justify-content-xs-space-between-30"
  >
    <div
      class="MuiGrid-root-10 makeStyles-messageLinkContainer-2 MuiGrid-container-11 MuiGrid-item-12 MuiGrid-align-items-xs-center-19 MuiGrid-grid-xs-true-44"
    >
      <div
        class="MuiGrid-root-10 MuiGrid-item-12"
      >
        <p
          class="MuiTypography-root-113 makeStyles-message-7 MuiTypography-body1-115"
        >
          This is an important message
        </p>
      </div>
      <div
        class="MuiGrid-root-10 MuiGrid-item-12"
      >
        <a
          class="MuiTypography-root-113 MuiLink-root-143 MuiLink-underlineHover-145 makeStyles-link-6 MuiTypography-colorPrimary-136"
          href="https://badger.com"
          rel="noreferrer"
          target="_blank"
        >
          <img
            alt="banner link icon"
            class="makeStyles-linkIcon-3"
            src="/assets/icons/banner-link-icon.svg"
          />
          view more
        </a>
      </div>
    </div>
  </div>
</div>
`;
